File: src/main/java/com/ptaf/ai/validation/PageFrameContextGuard.java
package com.ptaf.ai.validation;

import com.ptaf.ai.model.AiGenerationStructuredResponse;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class PageFrameContextGuard {
    private static final Pattern PAGE_FRAME_STEP = Pattern.compile(
            ".*on\\s+(page|frame)\\s+([a-zA-Z0-9_.-]+)\\s+(?:locator|of\\s+locator)\\s+([a-zA-Z0-9_.-]+).*",
            Pattern.CASE_INSENSITIVE
    );
    private static final Pattern UI_CONTEXT_HINT = Pattern.compile(".*on\\s+(page|frame)\\b.*", Pattern.CASE_INSENSITIVE);

    public PageFrameContextGuardResult validate(
            AiGenerationStructuredResponse structuredResponse,
            String defaultUiContext,
            List<String> frameAllowedPages,
            List<String> frameAllowedLocators
    ) {
        Set<String> allowedPages = normalizeSet(frameAllowedPages);
        Set<String> allowedLocators = normalizeSet(frameAllowedLocators);
        String featureFile = structuredResponse != null ? structuredResponse.featureFile() : "";

        List<String> invalidFrameSteps = new ArrayList<>();
        List<String> invalidPageSteps = new ArrayList<>();
        List<String> warnings = new ArrayList<>();
        List<String> blockingErrors = new ArrayList<>();
        int frameCount = 0;
        int pageCount = 0;

        for (String rawLine : featureFile.split("\\R")) {
            String line = rawLine.trim();
            if (line.isBlank()) {
                continue;
            }
            Matcher matcher = PAGE_FRAME_STEP.matcher(line);
            if (!matcher.matches()) {
                if (UI_CONTEXT_HINT.matcher(line).matches() && line.toLowerCase(Locale.ROOT).contains(" on frame ")) {
                    warnings.add("Could not parse frame step context cleanly: " + line);
                }
                continue;
            }

            String context = matcher.group(1).toLowerCase(Locale.ROOT);
            String page = matcher.group(2).toLowerCase(Locale.ROOT);
            String locator = matcher.group(3).toLowerCase(Locale.ROOT);

            if ("frame".equals(context)) {
                frameCount++;
                String compound = page + "." + locator;
                boolean allowed = allowedPages.contains(page) || allowedLocators.contains(compound);
                if (!allowed) {
                    invalidFrameSteps.add(line);
                    blockingErrors.add("Frame step is not allowed for page '" + page + "' locator '" + locator + "'. Use page step instead.");
                }
            } else {
                pageCount++;
                if (!"page".equals(defaultUiContext)) {
                    invalidPageSteps.add(line);
                    warnings.add("Encountered page step while defaultUiContext is '" + defaultUiContext + "': " + line);
                }
            }
        }

        boolean passed = blockingErrors.isEmpty();
        return new PageFrameContextGuardResult(
                passed,
                List.copyOf(invalidFrameSteps),
                List.copyOf(invalidPageSteps),
                List.copyOf(warnings),
                List.copyOf(blockingErrors),
                frameCount,
                pageCount
        );
    }

    private Set<String> normalizeSet(List<String> values) {
        if (values == null || values.isEmpty()) {
            return Set.of();
        }
        Set<String> out = new LinkedHashSet<>();
        for (String value : values) {
            if (value == null || value.isBlank()) {
                continue;
            }
            out.add(value.trim().toLowerCase(Locale.ROOT));
        }
        return out;
    }
}

File: ptaf/ai/validation/page_frame_context_guard.py
from __future__ import annotations

import re

from ptaf.ai.model.ai_generation_structured_response import AiGenerationStructuredResponse
from ptaf.ai.validation.page_frame_context_guard_result import PageFrameContextGuardResult

_PAGE_FRAME_STEP = re.compile(
    r".*on\s+(page|frame)\s+([a-zA-Z0-9_.-]+)\s+(?:locator|of\s+locator)\s+([a-zA-Z0-9_.-]+).*",
    re.IGNORECASE,
)
_UI_CONTEXT_HINT = re.compile(r".*on\s+(page|frame)\b.*", re.IGNORECASE)


class PageFrameContextGuard:
    def validate(
        self,
        structured_response: AiGenerationStructuredResponse,
        default_ui_context: str,
        frame_allowed_pages: list[str],
        frame_allowed_locators: list[str],
    ) -> PageFrameContextGuardResult:
        allowed_pages = _normalize_set(frame_allowed_pages)
        allowed_locators = _normalize_set(frame_allowed_locators)
        feature_file = structured_response.feature_file if structured_response else ""

        invalid_frame_steps: list[str] = []
        invalid_page_steps: list[str] = []
        warnings: list[str] = []
        blocking_errors: list[str] = []
        frame_count = 0
        page_count = 0

        for raw_line in feature_file.splitlines():
            line = raw_line.strip()
            if not line:
                continue
            match = _PAGE_FRAME_STEP.match(line)
            if not match:
                if _UI_CONTEXT_HINT.match(line) and " on frame " in line.lower():
                    warnings.append(f"Could not parse frame step context cleanly: {line}")
                continue

            context = match.group(1).lower()
            page = match.group(2).lower()
            locator = match.group(3).lower()

            if context == "frame":
                frame_count += 1
                compound = f"{page}.{locator}"
                allowed = page in allowed_pages or compound in allowed_locators
                if not allowed:
                    invalid_frame_steps.append(line)
                    blocking_errors.append(
                        f"Frame step is not allowed for page '{page}' locator '{locator}'. Use page step instead."
                    )
            else:
                page_count += 1
                if default_ui_context != "page":
                    invalid_page_steps.append(line)
                    warnings.append(
                        f"Encountered page step while defaultUiContext is '{default_ui_context}': {line}"
                    )

        return PageFrameContextGuardResult(
            passed=not blocking_errors,
            invalid_frame_steps=list(invalid_frame_steps),
            invalid_page_steps=list(invalid_page_steps),
            warnings=list(warnings),
            blocking_errors=list(blocking_errors),
            frame_step_count=frame_count,
            page_step_count=page_count,
        )


def _normalize_set(values: list[str] | None) -> set[str]:
    if not values:
        return set()
    out: set[str] = set()
    for value in values:
        if value and value.strip():
            out.add(value.strip().lower())
    return out
